File: units_cpp/src/unitsTest/cpp/constants/test_sun_properties.cc
/*
 * test_sun_properties.cc
 *
 *  Created on: Dec 26, 2017
 *      Author: sagis
 */

#include <units/constants/sun_properties.h>
#include <units/constants/physical_constants.h>
#include <units/units_multiply.h>
#include <units/units_ratio.h>
#include <units/units_pow.h>
#include <units/units_sqrt.h>
#include <units/planetary_constants_test_suite.h>

using namespace units::sun_properties;
using namespace units;
using units::physical_constants::gravitationalConstant;

class TestSunProperties : public PlanetaryConstantsTestSuite
{
protected:
	TestSunProperties()
	: PlanetaryConstantsTestSuite{0.14} // %
	{
	}
};

TEST_F(TestSunProperties, test_surface_gravity)
{
	auto calculatedSurfaceGravity =
			units_ratio<units::acceleration::MetersPerSquareSecond>(
					auto_units_multiply(gravitationalConstant, sunMass),
					auto_units_multiply(sunEquatorialRadius, sunEquatorialRadius));
	checkEqualValue(sunSurfaceGravity, calculatedSurfaceGravity);
}

TEST_F(TestSunProperties, test_mass)
{
	auto calculatedMass = (M_PI * 4 / 3) *
			units_multiply<units::mass::Kilograms>(
					sunMeanDensity ,units_pow<3>(sunEquatorialRadius));
	checkEqualValue(sunMass, calculatedMass);
}

TEST_F(TestSunProperties, test_escape_velocity)
{
	auto calculatedEscapeVelocity = units_sqrt<units::speed::MetersPerSecond> (
			2 * auto_units_ratio(
					auto_units_multiply(gravitationalConstant, sunMass),
					sunEquatorialRadius
				)
			);
	checkEqualValue(sunEscapeVelocity, calculatedEscapeVelocity);
}

TEST_F(TestSunProperties, test_solar_mass)
{
	checkEqualValue(units_cast<mass::SolarMasses>(sunMass), mass::SolarMasses{1});
}

File: units_java/src/test/java/com/units/tests/constants/TestSunProperties.java
package com.units.tests.constants;

import static com.units.constants.PhysicalConstants.gravitationalConstant;
import static com.units.constants.SunProperties.*;

import org.junit.Test;

import com.units.acceleration.MetersPerSquareSecond;
import com.units.mass.Kilograms;
import com.units.mass.SolarMasses;
import com.units.speed.MetersPerSecond;

public class TestSunProperties extends PlanetaryConstantsTestSuite {
	
	public TestSunProperties() {
		super(0.14); // %
	}
	@Test
	public void testSurfaceGravity() {
		MetersPerSquareSecond calculatedAcceleration =
				gravitationalConstant
				.multiply(sunMass)
				.divide(sunEquatorialRadius)
				.divide(sunEquatorialRadius)
				.castTo(MetersPerSquareSecond.class);
		checkAcceleration(sunSurfaceGravity, calculatedAcceleration);	
	}
	
	@Test
	public void testMass() {
		Kilograms calculatedMass =
				sunMeanDensity
				.multiply(sunEquatorialRadius.pow(3))
				.castTo(Kilograms.class)
				.multiplyByScalar(Math.PI * 4 / 3);
		checkMass(sunMass, calculatedMass);
	}
	
	@Test
	public void testEscapeVelocity() {
		MetersPerSecond calculatedEscapeVelocity =
				gravitationalConstant
				.multiply(sunMass)
				.divide(sunEquatorialRadius)
				.sqrt()
				.castTo(MetersPerSecond.class)
				.multiplyByScalar(Math.sqrt(2));
		checkVelocity(sunEscapeVelocity, calculatedEscapeVelocity);
	}
	
	@Test
	public void testSolarMass() {
		checkMass(sunMass, SolarMasses.one().castTo(Kilograms.class));
	}

}
